### ISSUE TITLE
V100 PCIE FP16 PS/worker training data corruption bug occurs when distributed over gRPC

### ISSUE THREAD

**Issue by byronyi** [tensorflow/tensorflow#23606]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): [tensorflow/benchmarks:cnn_tf_v1.12_compatible](https://github.com/tensorflow/benchmarks/tree/cnn_tf_v1.12_compatible)
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 16.04-hwe
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): ('v1.12.0-0-ga6d8ffae09', '1.12.0')
- Python version: 2.7.12
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: 9.0/7.2.1.38-1+cuda9.0
- GPU model and memory: V100-PCIE-32GB, NVIDIA driver version 384.145, with `intel_iommu=on`

```
$ cat /proc/cmdline
BOOT_IMAGE=/boot/vmlinuz-4.15.0-38-generic root=/dev/mapper/vgroot-lvroot ro intel_iommu=on
```
```
$ nvidia-smi topo -m
	GPU0	GPU1	GPU2	GPU3	CPU Affinity
GPU0	 X 	PIX	NODE	NODE	0-9,20-29
GPU1	PIX	 X 	NODE	NODE	0-9,20-29
GPU2	NODE	NODE	 X 	PIX	0-9,20-29
GPU3	NODE	NODE	PIX	 X 	0-9,20-29

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe switches (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing a single PCIe switch
  NV#  = Connection traversing a bonded set of # NVLinks
```

**Describe the current behavior**
FP-16 multi-GPU training with CPU as local parameter server is converging in single process mode, but diverging loss value (nan) adding another loopback PS process with grpc. Consistent behaviours for both ResNet-50 and VGG16. 

**Describe the expected behavior**
The same training procedure should yield exactly same result using only 1 worker with/without 1 PS. Not sure why adding SendRecvOps causes data corruption. See other info below.

**Code to reproduce the issue**

ResNet50 local with [output](https://gist.github.com/byronyi/4f0821a4792b479bf884ff81171c4011):
```
$ python benchmarks/scripts/tf_cnn_benchmarks/tf_cnn_benchmarks.py \
  --variable_update=parameter_server \
  --local_parameter_device=cpu \
  --model=resnet50 \
  --num_gpus=4 \
  --use_fp16 \
  --batch_size=256
```

VGG16 local with [output](https://gist.github.com/byronyi/0dc379b8feac762bc8d04cab2b240060):
```
$ python benchmarks/scripts/tf_cnn_benchmarks/tf_cnn_benchmarks.py \
  --variable_update=parameter_server \
  --local_parameter_device=cpu \
  --model=vgg16 \
  --num_gpus=4 \
  --use_fp16 \
  --batch_size=256
```

Distributed using the same PS command with [output](https://gist.github.com/byronyi/ea508b150d1e55b0ac41502c48685b7c):
```
CUDA_VISIBLE_DEVICES= \
python benchmarks/scripts/tf_cnn_benchmarks/tf_cnn_benchmarks.py \
  --ps_hosts=localhost:5000 \
  --worker_hosts=localhost:5001 \
  --job_name=ps \
  --local_parameter_device=cpu \
  --task_index=0 \
  --server_protocol=grpc
```

ResNet50 distributed with [output](https://gist.github.com/byronyi/de05ccb7b481932ce7be91f889f41aee):
```
$ python benchmarks/scripts/tf_cnn_benchmarks/tf_cnn_benchmarks.py \
  --ps_hosts=localhost:5000 \
  --worker_hosts=localhost:5001 \
  --job_name=worker \
  --task_index=0 \
  --server_protocol=grpc \
  --variable_update=parameter_server \
  --local_parameter_device=cpu \
  --model=resnet50 \
  --num_gpus=4 \
  --use_fp16 \
  --batch_size=256
```

VGG16 distributed with [output](https://gist.github.com/byronyi/1113a5cd8048f729a2c73df5bfd97017):
```
$ python benchmarks/scripts/tf_cnn_benchmarks/tf_cnn_benchmarks.py \
  --ps_hosts=localhost:5000 \
  --worker_hosts=localhost:5001 \
  --job_name=worker \
  --task_index=0 \
  --server_protocol=grpc \
  --variable_update=parameter_server \
  --local_parameter_device=cpu \
  --model=vgg16 \
  --num_gpus=4 \
  --use_fp16 \
  --batch_size=256
```

**Other info / logs**
Kernel reports `[DMA Write] PTE Write access is not set` for one of the GPU.
Detailed log could be found [here](https://gist.github.com/dd1e75702116d941d95dfa1fe17b8e27).

**Comment by byronyi** [comment]
I will try on one of our NVLink machines to see if the problem remains.

**Comment by byronyi** [comment]
I will try to reproduce this problem on one of our NVLink V100-SXM2 machine.

**Comment by byronyi** [comment]
The erroneous device has non-identity DMAR mapping, as shown in the kernel log:
``DMAR: 32bit 0000:1b:00.0 uses non-identity mapping``

**Comment by byronyi** [comment]
Disabling IOMMU temporarily resolves this issue. 

I am still wondering though, as it seems distributed runtime triggers GPU P2P transfer even with CPU as local PS, but the same error does not occur in local training. Is this behaviour intended? 

@poxvoculi Paul, is there anything you feel similar here?

**Comment by poxvoculi** [comment]
Byron,  we'll try to take a look at this before long, to see whether we can reproduce.

What do you mean by "it seems distributed runtime triggers GPU P2P transfer even with CPU as local PS"?  Are you seeing memcpy's direct between GPUs?  A lot of strange optimizations have been done to the TF CNN benchmark suite, so it's possible some kind of cross-device ops are intentionally done even in the CPU-as-ps case.

**Comment by byronyi** [comment]
Patching the latest firmware resolves the issue completely. I will leave it here and hope it could be helpful if someone else meet the same problem.

**Comment by MartinSmeyer** [comment]
Thanks a lot Byronyi! We will try this for our 8x Titan V, SuperMicro 4029GP-TRT Server. It shows very similar issues with the PIX connections.

**Comment by MartinSmeyer** [comment]
And it worked! :)
